src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -69,88 +31,34 @@ let internal rangeLiteralToSeq (r: Expr): seq<int64> =
-        | InvalidItem -> SingleItem "__invalid__"
-        | VariableName n -> SingleItem n.Value
-        | VariableNameTuple items -> MultipleItems (Seq.map StringTuple.fromSymbolTuple items)
-        | DiscardedItem -> SingleItem "__discarded__"
-
-    override this.ToString() =
-        match this with
-        | SingleItem item -> item
-        | MultipleItems items -> "(" + String.Join(", ", items) + ")"
-
-
-/// Modifies the VariablesDict by setting the given argument tuple to the given values
-let rec internal fillVars (vars: VariablesDict) (argTuple: StringTuple, arg: Expr): unit =
-    match argTuple with
-    | SingleItem item -> vars.defineVar(item, arg)
-    | MultipleItems items ->
-        match arg with
-        | ValueTuple vt ->
-            QsCompilerError.Verify (
-                vt.Length = Seq.length items,
-                "Matched tuples must have the same length")
-            vt |> Seq.map (fun x -> x.Expression) |>
-            Seq.zip items |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
-        | _ ->
-            QsCompilerError.Verify (
-                1 = Seq.length items,
-                "Expecting a single-element tuple")
-            Seq.zip items [arg] |> Seq.map (fillVars vars) |> List.ofSeq |> ignore
+/// Converts a QsTuple to a SymbolTuple
+let rec internal toSymbolTuple (x: QsTuple<LocalVariableDeclaration<QsLocalSymbol>>): SymbolTuple =
+    match x with
+    | QsTupleItem item ->
+        match item.VariableName with
+        | ValidName n -> VariableName n
+        | InvalidName -> InvalidItem
+    | QsTuple items when items.Length = 1 ->
+        toSymbolTuple items.[0]
+    | QsTuple items ->
+        VariableNameTuple ((Seq.map toSymbolTuple items).ToImmutableArray())
+
+/// Wraps a QsExpressionType in a basic TypedExpression
+/// The returned TypedExpression has no type param / inferred info / range information,
+/// and it should not be used for any code step that requires this information.
+let internal wrapExpr (expr: Expr) (bt: TypeKind): TypedExpression =
+    let ii = {IsMutable=false; HasLocalQuantumDependency=false}
+    TypedExpression.New (expr, ImmutableDictionary.Empty, ResolvedType.New bt, ii, Null)
+
+/// Wraps a QsStatementKind in a basic QsStatement
+let internal wrapStmt (stmt: QsStatementKind): QsStatement =
+    QsStatement.New QsComments.Empty Null (stmt, [])
 
 
 /// Returns a new array of the given type and length.
 /// Returns None if the type doesn't have a default value.
 let rec internal constructNewArray (bt: TypeKind) (length: int): Expr option =
-    match defaultValue bt with
-    | Some x -> ImmutableArray.CreateRange (List.replicate length (wrapExpr x bt)) |> ValueArray |> Some
-    | None -> None
+    defaultValue bt |> Option.map (fun x -> ImmutableArray.CreateRange (List.replicate length (wrapExpr x bt)) |> ValueArray)
 
 /// Returns the default value for a given type (from Q# documentation)
 and private defaultValue (bt: TypeKind): Expr option =
```

Comment:
What happens in the None case?

Reply:
It returns None, because it failed to create the array

Reply:
But why? You should be able to always create an array of known length, independent on the item type.

Reply:
For simplicity, I don't attempt to create arrays that contain invalid expressions, such as an uninitialized function or qubit